### PR TITLE
MINOR: Disable JDK 11 and 17 tests on PRs

### PR DESCRIPTION
**Pull request by gharris1727**
Testing these JDK versions in PRs is low value, and has a high cost in CI workload. At least temporarily we should disable these, and only run the full JDK suites for branch builds.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jolshan** [comment]
I was just having a conversation about this with some folks!

**Comment by mumrah** [comment]
I think it would be worth running the compile steps for these older versions (i.e., "doValidation"). As far as I know, those steps only fail if there's actually a problem.

File: Jenkinsfile
Comment (ijuma) [inline Jenkinsfile:43]: If they don't run on PRs, I don't think anyone will validate the results for the branch builders. We should probably remove Java 11 and 17 altogether now that we have ported most tests to run with Java 16 and higher (previously many were excluded).

I had suggested this in an internal conversation, so good to see similar action here.

**Comment by jolshan** [comment]
> I think it would be worth running the compile steps for these older versions (i.e., "doValidation"). As far as I know, those steps only fail if there's actually a problem.

It's rare that there is an issue with only java 11/17 and not 8 or 21. I've most frequently seen issues with scala 2.12 and that's on the java 8 build. If we still do the compile step, we still need the nodes right?

**Comment by gharris1727** [comment]
> It's rare that there is an issue with only java 11/17 and not 8 or 21. I've most frequently seen issues with scala 2.12 and that's on the java 8 build. 

I agree, and that was why I picked 11 and 17 to affect here.

> If we still do the compile step, we still need the nodes right?

We would still need the nodes, but instead of 4 nodes for 3 hours each (12 node-hours) we would need 2 nodes for 3 hours and 2 nodes for ~15 minutes (6.5 node-hours).

I'm not very familiar with the jenkins pipeline declaration syntax, and I'm not sure how to express this change in a way which doesn't generate 11 & 17 builds for PRs at all, because it seems like it can only evaluate conditions like isChangeRequest after the stage is already defined. Maybe someone that is more familiar with jenkins can suggest an alternative.

File: Jenkinsfile
Comment (gharris1727) [reply Jenkinsfile:43]: What do you think about leaving the 11 and 17 branch builders as-is until 4.0? This will need some rework then to effect the KIPs that are changing the minimum version requirements and having different minimums for the different modules. We can figure out a strategy for branch & pr builds then.

Comment (ijuma) [reply Jenkinsfile:43]: I don't see the value - once we drop support for Java 8, we will simply change the Java 8 builder to Java 11. Am I missing something?

Comment (ijuma) [reply Jenkinsfile:43]: Ah, yes - we'll also have to configure the brokers, etc. to use Java 17. In any case, I think whatever we have to do there won't be any harder if we just delete the Java 11 and 17 builders for now. And it will make the 3.8 builds lighter.

**Comment by gharris1727** [comment]
Looking at the latest CI run, the 11 build took ~15 minutes, the 17 build took 11 minutes, and the 8 and 21 builds are continuing as normal.

In the last 28 days, Kafka has run:
9380 builds in total
8514 builds on PRs
748 builds on trunk
111 builds on 3.7
7 builds on 3.6
-> 90% of the builds are on PRs

I think we can expect more than a 40% reduction in our total CI load after this change, and I don't think we're losing any test coverage. I'm comfortable with this change as-is, and we can pursue further reductions as they become desired.

File: Jenkinsfile
Comment (gharris1727) [reply Jenkinsfile:43]: I'm hesitant to sign up future release managers for additional verification of the 11 and 17 builds that the CI has been doing for previous releases. I don't expect any failures to be introduced that affect 11 and 17 but not 8 or 21, but I would rather find those failures out in CI after merge than during the RC generation stage.

And for selfish reasons, I also would like to continue seeing the 11 and 17 trunk results published so I can do flaky test analysis. With this patch as-is, we're keeping the same number of test runs on trunk, but if we removed these entirely we would be cutting those in half. I don't pay attention to the PR builds for statistics because contributors may have bad patches applied that aren't worth investigating.

**Comment by jolshan** [comment]
> Something is wrong with the build of this commit

There still seems to be some issue with builds 🤔

**Comment by jolshan** [comment]
Any ideas about 
```

[2024-05-23T21:45:16.078Z] > Task :connect:runtime:test

[2024-05-23T21:45:16.078Z] Could not stop org.gradle.internal.actor.internal.DefaultActorFactory$NonBlockingActor@1d0918ec.

[2024-05-23T21:45:16.078Z] org.gradle.internal.dispatch.DispatchException: Could not dispatch message [MethodInvocation method: stop()].
```

**Comment by gharris1727** [comment]
> > Something is wrong with the build of this commit
>
> There still seems to be some issue with builds 🤔

> [2024-05-23T21:45:16.078Z] Could not stop org.gradle.internal.actor.internal.DefaultActorFactory$NonBlockingActor@1d0918ec.

That was probably me. Marking the PR as ready-for-review forcefully killed the previous build, and the next build is still waiting in the queue. I think that the status message should say "queued", and it just wasn't prepared for this strange state. If it doesn't resolve in a few hours then we might want to be worried :)

**Comment by gharris1727** [comment]
Hey @jolshan @mumrah @ijuma PTAL at the build to see the effect, thanks!

**Review by jolshan — APPROVED** [review]
Seems reasonable to me. I don't have a strong opinion about if we drop 8 and 11 altogether or leave them for branch builds. We may want to hear back from Ismael on that though.

**Comment by ijuma** [comment]
I'm ok going with something like this to start with.

**Review by showuon — APPROVED** [review]
I like this proposal! After merging it and making sure the trunk branch still work as expect, we should raise this change in dev mailing list. Thanks @gharris1727 for this improvement!

**Comment by chia7712** [comment]
@gharris1727 nice patch. One question: Can we increase the parallel number from 2 to 4 after we disable two suits (11 and 17) of tests.

**Comment by gharris1727** [comment]
> Can we increase the parallel number from 2 to 4

I think that is a separate concern. Before or after this change, each "stage" in the Jenkins build gets either a full or half node, and so has the same resource limitations. What this changes is how long we reserve the half-node, and how soon another stage (of another Kafka build, or another project's build) can start running.

I still need to run some more experiments in #16009 to see if we need to change maxParallelForks.

**Pull request merged**
